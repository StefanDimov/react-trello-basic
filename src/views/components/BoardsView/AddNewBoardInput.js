import React from 'react'
import { FormGroup, FormControl } from 'react-bootstrap'

/**
 * Component with an input for adding a new board. Mean to be used in BoardItemsList.
 * @class comp/AddNewBoardInput
 * @param {function} onCreateBoard - callback for on input submit
 */
export default class AddNewBoardInput extends React.Component {
    constructor(props) { // eslint-disable-line
        super(props)

        this.state = { value: '' }

        this.changeValueHandler = this.changeValueHandler.bind(this)
        this.onSubmit = this.onSubmit.bind(this)
    }

    /**
     * Change handler set on the input
     * @private
     * @param {object} e event object provided to handler
     */
    changeValueHandler(e) {
        this.setState({ value: e.target.value })
    }

    /**
     * Form submit handler. Calls callback with input value.
     * @private
     * @param {object} e event object provided to handler
     */
    onSubmit(e) {
        e.preventDefault()
        this.props.onCreateBoard(this.state.value)
        this.setState({ value: '' })
    }

    render() { // eslint-disable-line
        return (
            <form className="my-add-new-board" onSubmit={this.onSubmit}>
                <FormGroup>
                    <FormControl
                        type="text"
                        value={this.state.value}
                        onChange={this.changeValueHandler}
                        placeholder="Add new board..." />
                </FormGroup>
            </form>
        )
    }
}

AddNewBoardInput.propTypes = {
    onCreateBoard: React.PropTypes.func.isRequired
}
