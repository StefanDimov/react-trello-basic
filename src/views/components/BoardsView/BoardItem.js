import React from 'react'
import * as boardUtils from '../../../utils/boardUtils'
import { DropdownButton, MenuItem } from 'react-bootstrap'

const DELETE_EVENT_KEY = 1

/**
 * Component that represents a Board Item. Ment to be used with BoardItemsList Component
 * @class comp/BoardItem
 * @param {object} board - The Board data
 * @param {function} onDeleteBoard - function to execute when the board should be deleted
 */
export default class BoardItem extends React.Component {

    constructor(props) { // eslint-disable-line
        super(props)

        this.preventEventPropagation = this.preventEventPropagation.bind(this)
        this.onActionSelect = this.onActionSelect.bind(this)
    }

    /**
     * Stops event propagation
     * @param {object} e the event object
     */
    preventEventPropagation(e) {
        e.stopPropagation()
    }

    /**
     * Handler for the dropdown actions
     * @param {number} eventKey the eventKey number of the MenuItem
     * @param {object} e the event obejct
     */
    onActionSelect(eventKey, e) {
        e.stopPropagation()

        switch (eventKey) {
        case DELETE_EVENT_KEY: // Delete option
            this.props.onDeleteBoard(this.props.board)
            break
        }
    }

    render() { // eslint-disable-line
        const { board } = this.props
        return (
            <div className="my-board-item">
                <div className="my-board-item-title">{ board.title }</div>
                <DropdownButton
                    onClick={this.preventEventPropagation}
                    onSelect={this.onActionSelect}
                    pullRight
                    className="my-board-item-dropdown-overrides"
                    bsStyle="link"
                    title=""
                    id={board.id + '-dropdown'}>
                        <MenuItem eventKey={DELETE_EVENT_KEY}>Delete</MenuItem>
                </DropdownButton>
            </div>
        )
    }
}

BoardItem.propTypes = {
    board: boardUtils.boardPropType.isRequired,
    onDeleteBoard: React.PropTypes.func.isRequired
}
