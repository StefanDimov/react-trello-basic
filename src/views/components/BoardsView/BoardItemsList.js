import React from 'react'

import { Row, Col } from 'react-bootstrap'

import * as boardUtils from '../../../utils/boardUtils'
import BoardItem from './BoardItem'
import AddNewBoardInput from './AddNewBoardInput'

/**
 * Component that displays BoardItems. Has an input for adding a new board.
 * @class comp/BoardItemsList
 * @param {array} boards Boards to display
 * @param {function} onBoardClick callback for when a board is clicked
 * @param {function} onCreateBoard callback with the data to from the new board input
 * @param {function} onDeleteBoard callback which will be called when a board should be deleted
 */
export default class BoardItemsList extends React.Component {
    render() { // eslint-disable-line
        const { boards, onBoardClick, onCreateBoard, onDeleteBoard } = this.props
        return (
            <Row>
                {boards.map((board) =>
                    <Col sm={3} key={board.id} onClick={() => onBoardClick(board)}>
                        <BoardItem board={board} onDeleteBoard={onDeleteBoard} />
                    </Col>
                )}
                <Col sm={3} key={'add-new-boards-item'}>
                    <AddNewBoardInput onCreateBoard={onCreateBoard} />
                </Col>
            </Row>
        )
    }
}

BoardItemsList.propTypes = {
    boards: React.PropTypes.arrayOf(boardUtils.boardPropType).isRequired,
    onBoardClick: React.PropTypes.func.isRequired,
    onCreateBoard: React.PropTypes.func.isRequired,
    onDeleteBoard: React.PropTypes.func.isRequired
}
