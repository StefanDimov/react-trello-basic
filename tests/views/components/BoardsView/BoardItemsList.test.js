import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import { getEmptyBoard } from '../../../_mocks/Board.mocks.js'

import BoardItemsList from '../../../../src/views/components/BoardsView/BoardItemsList'
import BoardItem from '../../../../src/views/components/BoardsView/BoardItem'
import AddNewBoardInput from '../../../../src/views/components/BoardsView/AddNewBoardInput'

describe('BoardItemsList', () => {

    const onBoardClick = jest.fn()
    const onCreateBoard = jest.fn()
    const onDeleteBoard = jest.fn()
    let board, anotherBoard, thirdBoard

    beforeEach(() => {
        board = getEmptyBoard(1)
        anotherBoard = getEmptyBoard(2)
        thirdBoard = getEmptyBoard(3)

        onBoardClick.mockClear()
        onCreateBoard.mockClear()
        onDeleteBoard.mockClear()
    })

    describe('rendering', () => {

        it('should render properly with no boards', () => {
            const tree = renderer.create(
                <BoardItemsList boards={[]}
                    onBoardClick={onBoardClick}
                    onCreateBoard={onCreateBoard}
                    onDeleteBoard={onDeleteBoard} />
            ).toJSON()

            expect(tree).toMatchSnapshot()
        })

        it('should render properly with one board', () => {
            const tree = renderer.create(
                <BoardItemsList boards={[board]}
                    onBoardClick={onBoardClick}
                    onCreateBoard={onCreateBoard}
                    onDeleteBoard={onDeleteBoard} />
            ).toJSON()

            expect(tree).toMatchSnapshot()
        })

        it('should render properly with multiple boards', () => {
            const tree = renderer.create(
                <BoardItemsList boards={[board, anotherBoard, thirdBoard]}
                    onBoardClick={onBoardClick}
                    onCreateBoard={onCreateBoard}
                    onDeleteBoard={onDeleteBoard} />
            ).toJSON()

            expect(tree).toMatchSnapshot()
        })
    })

    describe('onBoardClick', () => {
        it('should be set on all BoardItems container', () => {
            const wrapper = shallow(
                <BoardItemsList boards={[board, anotherBoard]}
                    onBoardClick={onBoardClick}
                    onCreateBoard={onCreateBoard}
                    onDeleteBoard={onDeleteBoard} />
            )

            const firstContainer = wrapper.find(BoardItem).first().parent()
            const secondContainer = wrapper.find(BoardItem).last().parent()

            expect(firstContainer.prop('onClick')).toBeDefined()
            expect(secondContainer.prop('onClick')).toBeDefined()
        })

        it('should be called with appropriate params on click', () => {
            const wrapper = shallow(
                <BoardItemsList boards={[board]}
                    onBoardClick={onBoardClick}
                    onCreateBoard={onCreateBoard}
                    onDeleteBoard={onDeleteBoard} />
            )

            const container = wrapper.find(BoardItem).parent()
            container.simulate('click')

            expect(onBoardClick).toHaveBeenCalledTimes(1)
            expect(onBoardClick).toHaveBeenCalledWith(board)
        })
    })

    describe('onCreateBoard', () => {
        it('should be set on AddNewBoardInput', () => {
            const wrapper = shallow(
                <BoardItemsList boards={[board]}
                    onBoardClick={onBoardClick}
                    onCreateBoard={onCreateBoard}
                    onDeleteBoard={onDeleteBoard} />
            )
            expect(wrapper.find(AddNewBoardInput).prop('onCreateBoard')).toBe(onCreateBoard)
        })
    })

    describe('onDeleteBoard', () => {
        it('should be set each rendered BoardItem', () => {
            const wrapper = shallow(
                <BoardItemsList boards={[board, anotherBoard]}
                    onBoardClick={onBoardClick}
                    onCreateBoard={onCreateBoard}
                    onDeleteBoard={onDeleteBoard} />
            )
            expect(wrapper.find(BoardItem).first().prop('onDeleteBoard')).toBe(onDeleteBoard)
            expect(wrapper.find(BoardItem).last().prop('onDeleteBoard')).toBe(onDeleteBoard)
        })
    })
})
