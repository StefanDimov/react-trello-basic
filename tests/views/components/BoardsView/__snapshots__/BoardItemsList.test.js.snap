exports[`BoardItemsList rendering should render properly with multiple boards 1`] = `
<div
  className="row">
  <div
    className="col-sm-3"
    onClick={[Function]}>
    <div
      className="my-board-item">
      <div
        className="my-board-item-title">
        boardTitle1
      </div>
      <div
        className="dropdown btn-group btn-group-link">
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="my-board-item-dropdown-overrides dropdown-toggle btn btn-link"
          disabled={false}
          id="boardId1-dropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button">
           
          <span
            className="caret" />
        </button>
        <ul
          aria-labelledby="boardId1-dropdown"
          className="dropdown-menu dropdown-menu-right"
          role="menu">
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              Delete
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="col-sm-3"
    onClick={[Function]}>
    <div
      className="my-board-item">
      <div
        className="my-board-item-title">
        boardTitle2
      </div>
      <div
        className="dropdown btn-group btn-group-link">
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="my-board-item-dropdown-overrides dropdown-toggle btn btn-link"
          disabled={false}
          id="boardId2-dropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button">
           
          <span
            className="caret" />
        </button>
        <ul
          aria-labelledby="boardId2-dropdown"
          className="dropdown-menu dropdown-menu-right"
          role="menu">
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              Delete
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="col-sm-3"
    onClick={[Function]}>
    <div
      className="my-board-item">
      <div
        className="my-board-item-title">
        boardTitle3
      </div>
      <div
        className="dropdown btn-group btn-group-link">
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="my-board-item-dropdown-overrides dropdown-toggle btn btn-link"
          disabled={false}
          id="boardId3-dropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button">
           
          <span
            className="caret" />
        </button>
        <ul
          aria-labelledby="boardId3-dropdown"
          className="dropdown-menu dropdown-menu-right"
          role="menu">
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              Delete
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="col-sm-3">
    <form
      className="my-add-new-board"
      onSubmit={[Function]}>
      <div
        className="form-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          placeholder="Add new board..."
          type="text"
          value="" />
      </div>
    </form>
  </div>
</div>
`;

exports[`BoardItemsList rendering should render properly with no boards 1`] = `
<div
  className="row">
  <div
    className="col-sm-3">
    <form
      className="my-add-new-board"
      onSubmit={[Function]}>
      <div
        className="form-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          placeholder="Add new board..."
          type="text"
          value="" />
      </div>
    </form>
  </div>
</div>
`;

exports[`BoardItemsList rendering should render properly with one board 1`] = `
<div
  className="row">
  <div
    className="col-sm-3"
    onClick={[Function]}>
    <div
      className="my-board-item">
      <div
        className="my-board-item-title">
        boardTitle1
      </div>
      <div
        className="dropdown btn-group btn-group-link">
        <button
          aria-expanded={false}
          aria-haspopup={true}
          className="my-board-item-dropdown-overrides dropdown-toggle btn btn-link"
          disabled={false}
          id="boardId1-dropdown"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          type="button">
           
          <span
            className="caret" />
        </button>
        <ul
          aria-labelledby="boardId1-dropdown"
          className="dropdown-menu dropdown-menu-right"
          role="menu">
          <li
            className=""
            role="presentation"
            style={undefined}>
            <a
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="menuitem"
              tabIndex="-1">
              Delete
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="col-sm-3">
    <form
      className="my-add-new-board"
      onSubmit={[Function]}>
      <div
        className="form-group">
        <input
          className="form-control"
          id={undefined}
          onChange={[Function]}
          placeholder="Add new board..."
          type="text"
          value="" />
      </div>
    </form>
  </div>
</div>
`;
